apk/apk.go

```diff
@@ -0,0 +1,597 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err = gw.Close(); err != nil {
+		return nil, err
+	}
+
+	return digest.Sum(nil), nil
+}
+
+func parseRsaPrivateKeyFromPemStr(privPEM string) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(privPEM))
+	if block == nil {
+		return nil, errors.New("failed to parse PEM block containing the key")
+	}
+
+	priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return priv, nil
+}
+
+func fileToBase64String(file string) (string, error) {
+	fileBytes, err := ioutil.ReadFile(filepath.Clean(file))
+	if err != nil {
+		return "", err
+	}
+	base64Contents := base64.StdEncoding.EncodeToString(fileBytes)
+	return base64Contents, nil
+}
+
+func createSignature(signatureTgz io.Writer, controlDigest []byte, info *nfpm.Info) error {
+	var pemBytes []byte
+	var err error
+	switch {
+	case info.PrivateKeyBase64 != "":
+		if pemBytes, err = base64.StdEncoding.DecodeString(info.PrivateKeyBase64); err != nil {
+			return err
+		}
+	case info.PrivateKeyFile != "":
+		if pemBytes, err = ioutil.ReadFile(filepath.Clean(info.PrivateKeyFile)); err != nil {
+			return err
+		}
+	default:
+		fmt.Printf("yaml configuration of 'privatekeybase64' or 'privatekeyfile' is required for .apk\n")
+	}
+	priv, err := parseRsaPrivateKeyFromPemStr(string(pemBytes))
+	if err != nil {
+		return err
+	}
+	signed, err := priv.Sign(rand.Reader, controlDigest, crypto.SHA256)
+	if err != nil {
+		return err
+	}
+	// fmt.Println("data sign  :", hex.EncodeToString(signed))
+
+	// create the signature tgz
+	builderSignature := createBuilderSignature(signed, err)
+	_, err = writeTgz(signatureTgz, tarCut, builderSignature, sha256.New())
+	return err
+}
+
+func createData(dataTgz io.Writer, info *nfpm.Info, sizep *int64) ([]byte, error) {
+	builderData := createBuilderData(info, sizep)
+	dataDigest, err := writeTgz(dataTgz, tarFull, builderData, sha256.New())
+	if err != nil {
+		return nil, err
+	}
+	// fmt.Println("sizep      :", *sizep)
+	// fmt.Println("data sha256:", hex.EncodeToString(dataDigest))
+	return dataDigest, nil
+}
+
+func createControl(controlTgz io.Writer, info *nfpm.Info, dataDigest []byte, size int64) ([]byte, error) {
+	builderControl := createBuilderControl(info, size, dataDigest)
+	controlDigest, err := writeTgz(controlTgz, tarCut, builderControl, sha256.New())
+	if err != nil {
+		return nil, err
+	}
+	// fmt.Println("data sha1  :", hex.EncodeToString(controlDigest))
+	return controlDigest, nil
+}
+
+func combineToApk(target io.Writer, dataTgz, controlTgz, signatureTgz io.Reader) (err error) {
+	tgzs := []io.Reader{signatureTgz, controlTgz, dataTgz}
+
+	for _, tgz := range tgzs {
+		if _, err = io.Copy(target, tgz); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func createBuilderSignature(signed []byte, err error) func(tw *tar.Writer) error {
+	return func(tw *tar.Writer) error {
+		keyname := "alpine-devel@lists.alpinelinux.org-4a6a0840"
```

Comment:
This should probably be configurable.

Reply:
> This should probably be configurable.

Agreed. Added a new setting at the root.  769b4ad